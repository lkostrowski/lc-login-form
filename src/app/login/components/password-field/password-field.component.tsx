import React, { useState } from 'react';
import cx from 'classnames';

import { ReactComponent as EyeIcon } from '../../../../icons/eye.svg';

import styles from './password-field.module.scss';

import {
    Props as TextFieldProps,
    TextField,
} from '../../../common/text-field/text-field.component';
import { ConditionBar } from '../../../common/condition-bar/condition-bar.component';
import { usePasswordValidator } from '../../hooks/use-password-validator';

interface Props extends TextFieldProps {}

const mapValidationToConditionState = (valid: boolean | undefined) =>
    valid === true ? 'valid' : valid === false ? 'invalid' : undefined;

export const PasswordField: React.FC<Props> = ({
    className,
    type,
    value,
    ...props
}) => {
    const [visible, setVisible] = useState(false);
    const { hasMinLength, hasUpper, hasLower, hasDigit } = usePasswordValidator(
        value as string,
    );

    return (
        <div className={className}>
            <TextField
                value={value}
                rightIcon={
                    <EyeIcon
                        className={cx(styles.icon, {
                            [styles.iconDisabled]: visible,
                        })}
                        onClick={() => setVisible((visible) => !visible)}
                    />
                }
                type={visible ? 'text' : 'password'}
                {...props}
            />
            {(value as string).length > 0 && (
                <div className={styles.conditions} data-testid="password-field:validation-conditions">
                    <ConditionBar
                        className={styles.condition}
                        label="Min. 6 letters"
                        state={mapValidationToConditionState(hasMinLength)}
                    />
                    <ConditionBar
                        className={styles.condition}
                        label="Upper letter"
                        state={mapValidationToConditionState(hasUpper)}
                    />
                    <ConditionBar
                        className={styles.condition}
                        label="Lower letter"
                        state={mapValidationToConditionState(hasLower)}
                    />
                    <ConditionBar
                        className={styles.condition}
                        label="A digit"
                        state={mapValidationToConditionState(hasDigit)}
                    />
                </div>
            )}
        </div>
    );
};
