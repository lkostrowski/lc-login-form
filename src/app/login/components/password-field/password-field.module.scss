.icon {
    transition: 0.2s;
    cursor: pointer;
}

.iconDisabled {
    opacity: 0.5;
}
