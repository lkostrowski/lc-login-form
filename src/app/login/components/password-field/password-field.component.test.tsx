import { cleanup, render } from '@testing-library/react';
import nanoID from 'nanoid';
import React from 'react';

import { PasswordField } from './password-field.component';

afterEach(cleanup);

jest.mock('nanoid');
(nanoID as jest.Mock).mockImplementation(() => 'id');

describe('Password field', function() {
    describe('Snapshots', () => {
        it('Valid password', () => {
            const { container } = render(
                <PasswordField value='Password1' onChange={() => {}} />,
            );

            expect(container).toMatchSnapshot();
        });

        it('Empty password', () => {
            const { container } = render(
                <PasswordField value='' onChange={() => {}} />,
            );

            expect(container).toMatchSnapshot();
        });

        it('Invalid password', () => {
            const { container } = render(
                <PasswordField value='pass' onChange={() => {}} />,
            );

            expect(container).toMatchSnapshot();
        });
    });
});
