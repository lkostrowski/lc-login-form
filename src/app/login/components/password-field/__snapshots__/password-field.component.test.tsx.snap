// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Password field Snapshots Empty password 1`] = `
<div>
  <div>
    <div
      class="container"
    >
      <div
        class="fieldContainer"
      >
        <input
          class="field withRightIcon"
          id="id"
          type="password"
          value=""
        />
        <div
          class="rightIconContainer"
        >
          <svg
            class="icon"
          >
            eye.svg
          </svg>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Password field Snapshots Invalid password 1`] = `
<div>
  <div>
    <div
      class="container"
    >
      <div
        class="fieldContainer"
      >
        <input
          class="field withRightIcon"
          id="id"
          type="password"
          value="pass"
        />
        <div
          class="rightIconContainer"
        >
          <svg
            class="icon"
          >
            eye.svg
          </svg>
        </div>
      </div>
    </div>
    <div
      class="conditions"
      data-testid="password-field:validation-conditions"
    >
      <div
        class="condition container"
      >
        <div
          class="bar invalid"
        />
        <span
          class="label invalid"
        >
          Min. 6 letters
        </span>
      </div>
      <div
        class="condition container"
      >
        <div
          class="bar invalid"
        />
        <span
          class="label invalid"
        >
          Upper letter
        </span>
      </div>
      <div
        class="condition container"
      >
        <div
          class="bar valid"
        />
        <span
          class="label valid"
        >
          Lower letter
        </span>
      </div>
      <div
        class="condition container"
      >
        <div
          class="bar invalid"
        />
        <span
          class="label invalid"
        >
          A digit
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`Password field Snapshots Valid password 1`] = `
<div>
  <div>
    <div
      class="container"
    >
      <div
        class="fieldContainer"
      >
        <input
          class="field withRightIcon"
          id="id"
          type="password"
          value="Password1"
        />
        <div
          class="rightIconContainer"
        >
          <svg
            class="icon"
          >
            eye.svg
          </svg>
        </div>
      </div>
    </div>
    <div
      class="conditions"
      data-testid="password-field:validation-conditions"
    >
      <div
        class="condition container"
      >
        <div
          class="bar valid"
        />
        <span
          class="label valid"
        >
          Min. 6 letters
        </span>
      </div>
      <div
        class="condition container"
      >
        <div
          class="bar valid"
        />
        <span
          class="label valid"
        >
          Upper letter
        </span>
      </div>
      <div
        class="condition container"
      >
        <div
          class="bar valid"
        />
        <span
          class="label valid"
        >
          Lower letter
        </span>
      </div>
      <div
        class="condition container"
      >
        <div
          class="bar valid"
        />
        <span
          class="label valid"
        >
          A digit
        </span>
      </div>
    </div>
  </div>
</div>
`;
